dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
-      );
+  String generateRulebook(String rulebookId) throws RulebookNotFoundException, RuleNotFoundException, IOException {
+    Collection<Field<?>> rulebookKeyFields = new ArrayList<>();
+    rulebookKeyFields.add(Fields.stringField(ID_COL, rulebookId));
+
+    Optional<StructuredRow> optionalRulebookStructuredRow = rulebookTable.read(rulebookKeyFields);
+
+    if (!optionalRulebookStructuredRow.isPresent()) {
+      throw new RulebookNotFoundException(String.format("Rulebook '%s' not found.", rulebookId));
     }
 
-    List<String> ruleSet = convertRulesToSet(row.getString(RULES));
+    StructuredRow row = optionalRulebookStructuredRow.get();
+    List<String> ruleSet = convertRulesToSet(row.getString(RULES_COL));
     List<String> ruleOutput = new ArrayList<>();
+
     for (String rule : ruleSet) {
-      Row ruleRow = rules.get(toKey(rule));
-      if (ruleRow.isEmpty()) {
+      Collection<Field<?>> ruleKeyFields = new ArrayList<>();
+      ruleKeyFields.add(Fields.stringField(ID_COL, rule));
+
+      Optional<StructuredRow> optionalRuleStructuredRow = rulesTable.read(ruleKeyFields);
+
+      if (!optionalRuleStructuredRow.isPresent()) {
         throw new RuleNotFoundException(
           String.format("Rulebook '%s' contains rule '%s', but rule '%s' is not present in rules database.",
                         rulebookId, rule, rule)
         );
       }
-      ruleOutput.add(String.format(RULE_TEMPLATE, ruleRow.getString(ID), ruleRow.getString(DESCRIPTION),
-                                   ruleRow.getString(CONDITION), ruleRow.getString(ACTION)));
+
+      StructuredRow ruleRow = optionalRuleStructuredRow.get();
+      ruleOutput.add(String.format(RULE_TEMPLATE, ruleRow.getString(ID_COL), ruleRow.getString(DESCRIPTION_COL),
+                                   ruleRow.getString(CONDITION_COL), ruleRow.getString(ACTION_COL)));
     }
 
-    String format = String.format(RULEBOOK_TEMPLATE, row.getString(ID), row.getLong(VERSION),
-                                  row.getString(DESCRIPTION), row.getLong(CREATED),
-                                  row.getLong(UPDATED), row.getString(SOURCE),
-                                  row.getString(USER), Joiner.on("").join(ruleOutput));
-    return format;
+    return String.format(RULEBOOK_TEMPLATE, row.getString(ID_COL), row.getLong(VERSION_COL),
+                         row.getString(DESCRIPTION_COL), row.getLong(CREATED_COL),
+                         row.getLong(UPDATED_COL), row.getString(SOURCE_COL),
+                         row.getString(USER_COL), Joiner.on("").join(ruleOutput));
   }
 
   /**
    * Lists the rules and rulebooks present in the system.
    *
    * @return List of rules in the system.
    */
-  public List<Map<String, Object>> rules() {
-    List<Map<String, Object>> result = new ArrayList<>();
-    try (Scanner scan = rules.scan(null, null)) {
-      Row next;
-      while ((next = scan.next()) != null) {
+  List<Map<String, Object>> rules() throws IOException {
```

Comment:
is this called by anything? This seems like a bad idea, doing a full table scan and storing everything in memory.

If it's not being called, let's remove it. If it is, please open a jira to fix this, probably by adding pagination to the API.

Reply:
Yes, this is called by the endpoint to list all of the rules.

Should I add a ticket in CDAP's jira?

Reply:
yes, please open a CDAP jira.

Reply:
JIRA: https://issues.cask.co/browse/CDAP-15695